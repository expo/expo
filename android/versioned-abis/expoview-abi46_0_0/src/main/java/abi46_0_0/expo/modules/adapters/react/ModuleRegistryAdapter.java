package abi46_0_0.expo.modules.adapters.react;

import abi46_0_0.com.facebook.react.ReactPackage;
import abi46_0_0.com.facebook.react.bridge.NativeModule;
import abi46_0_0.com.facebook.react.bridge.ReactApplicationContext;
import abi46_0_0.com.facebook.react.uimanager.ViewManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import abi46_0_0.expo.modules.adapters.react.views.SimpleViewManagerAdapter;
import abi46_0_0.expo.modules.adapters.react.views.ViewGroupManagerAdapter;
import abi46_0_0.expo.modules.core.ModuleRegistry;
import abi46_0_0.expo.modules.core.interfaces.InternalModule;
import abi46_0_0.expo.modules.core.interfaces.Package;
import abi46_0_0.expo.modules.kotlin.KotlinInteropModuleRegistry;
import abi46_0_0.expo.modules.kotlin.ModulesProvider;
import abi46_0_0.expo.modules.kotlin.views.ViewWrapperDelegateHolder;

/**
 * An adapter over {@link ModuleRegistry}, compatible with React (implementing {@link ReactPackage}).
 * Provides React Native with native modules and view managers,
 * which in turn are created by packages provided by {@link ReactModuleRegistryProvider}.
 */
public class ModuleRegistryAdapter implements ReactPackage {
  protected ReactModuleRegistryProvider mModuleRegistryProvider;
  protected ModulesProvider mModulesProvider;
  protected ReactAdapterPackage mReactAdapterPackage = new ReactAdapterPackage();
  private NativeModulesProxy mModulesProxy;
  // We need to save all view holders to update them when the new kotlin module registry will be created.
  private List<ViewWrapperDelegateHolder> mWrapperDelegateHolders = null;

  public ModuleRegistryAdapter(List<Package> packageList) {
    mModuleRegistryProvider = new ReactModuleRegistryProvider(packageList, null);
  }

  public ModuleRegistryAdapter(ReactModuleRegistryProvider moduleRegistryProvider) {
    mModuleRegistryProvider = moduleRegistryProvider;
  }

  public ModuleRegistryAdapter(ReactModuleRegistryProvider moduleRegistryProvider, ModulesProvider modulesProvider) {
    mModuleRegistryProvider = moduleRegistryProvider;
    mModulesProvider = modulesProvider;
  }

  @Override
  public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
    ModuleRegistry moduleRegistry = mModuleRegistryProvider.get(reactContext);

    for (InternalModule internalModule : mReactAdapterPackage.createInternalModules(reactContext)) {
      moduleRegistry.registerInternalModule(internalModule);
    }

    List<NativeModule> nativeModules = getNativeModulesFromModuleRegistry(reactContext, moduleRegistry);
    if (mWrapperDelegateHolders != null) {
      KotlinInteropModuleRegistry kotlinInteropModuleRegistry = mModulesProxy.getKotlinInteropModuleRegistry();
      kotlinInteropModuleRegistry.updateModuleHoldersInViewManagers(mWrapperDelegateHolders);
    }

    return nativeModules;
  }

  protected List<NativeModule> getNativeModulesFromModuleRegistry(ReactApplicationContext reactContext, ModuleRegistry moduleRegistry) {
    List<NativeModule> nativeModulesList = new ArrayList<>(2);

    mModulesProxy = createNativeModulesProxy(reactContext, moduleRegistry);
    nativeModulesList.add(mModulesProxy);

    // Add listener that will notify abi46_0_0.expo.modules.core.ModuleRegistry when all modules are ready
    nativeModulesList.add(new ModuleRegistryReadyNotifier(moduleRegistry));

    ReactPackagesProvider reactPackagesProvider = moduleRegistry.getModule(ReactPackagesProvider.class);
    for (ReactPackage reactPackage : reactPackagesProvider.getReactPackages()) {
      nativeModulesList.addAll(reactPackage.createNativeModules(reactContext));
    }

    return nativeModulesList;
  }

  @Override
  @SuppressWarnings("unchecked")
  public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
    List<ViewManager> viewManagerList = new ArrayList<>(mModuleRegistryProvider.getReactViewManagers(reactContext));

    for (abi46_0_0.expo.modules.core.ViewManager viewManager : mModuleRegistryProvider.getViewManagers(reactContext)) {
      switch (viewManager.getViewManagerType()) {
        case GROUP:
          viewManagerList.add(new ViewGroupManagerAdapter(viewManager));
          break;
        case SIMPLE:
          viewManagerList.add(new SimpleViewManagerAdapter(viewManager));
          break;
      }
    }

    // We assume that `createNativeModules` was called first.
    NativeModulesProxy modulesProxy = Objects.requireNonNull(mModulesProxy);
    KotlinInteropModuleRegistry kotlinInteropModuleRegistry = modulesProxy.getKotlinInteropModuleRegistry();
    List<ViewManager<?, ?>> kViewManager = kotlinInteropModuleRegistry.exportViewManagers();
    // Saves all holders that needs to be in sync with module registry
    mWrapperDelegateHolders = kotlinInteropModuleRegistry.extractViewManagersDelegateHolders(kViewManager);
    viewManagerList.addAll(kViewManager);

    return viewManagerList;
  }

  private NativeModulesProxy createNativeModulesProxy(ReactApplicationContext reactContext, ModuleRegistry moduleRegistry) {
    if (mModulesProvider != null) {
      return new NativeModulesProxy(reactContext, moduleRegistry, mModulesProvider);
    } else {
      return new NativeModulesProxy(reactContext, moduleRegistry);
    }
  }
}
