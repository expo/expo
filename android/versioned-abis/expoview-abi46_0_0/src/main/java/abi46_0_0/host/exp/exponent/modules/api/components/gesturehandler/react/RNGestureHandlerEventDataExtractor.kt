package abi46_0_0.host.exp.exponent.modules.api.components.gesturehandler.react

import abi46_0_0.com.facebook.react.bridge.WritableMap
import abi46_0_0.host.exp.exponent.modules.api.components.gesturehandler.GestureHandler

interface RNGestureHandlerEventDataExtractor<T : GestureHandler<T>> {
  fun extractEventData(handler: T, eventData: WritableMap)
}
