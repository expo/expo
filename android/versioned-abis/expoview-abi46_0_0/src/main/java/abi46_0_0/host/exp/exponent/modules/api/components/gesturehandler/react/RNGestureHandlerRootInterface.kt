package abi46_0_0.host.exp.exponent.modules.api.components.gesturehandler.react

interface RNGestureHandlerRootInterface {
  val rootHelper: RNGestureHandlerRootHelper?
}
