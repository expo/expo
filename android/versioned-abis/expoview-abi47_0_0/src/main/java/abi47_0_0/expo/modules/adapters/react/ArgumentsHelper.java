package abi47_0_0.expo.modules.adapters.react;

import abi47_0_0.com.facebook.react.bridge.Dynamic;

import abi47_0_0.expo.modules.core.arguments.MapArguments;
import abi47_0_0.expo.modules.core.arguments.ReadableArguments;

public class ArgumentsHelper {
  public static Object getNativeArgumentForExpectedClass(Dynamic argument, Class<?> expectedArgumentClass) {
    switch (argument.getType()) {
      case String:
        return argument.asString();
      case Map:
        if (expectedArgumentClass.isAssignableFrom(ReadableArguments.class)) {
          return new MapArguments(argument.asMap().toHashMap());
        }
        return argument.asMap().toHashMap();
      case Array:
        return argument.asArray().toArrayList();
      case Number:
        // Argument of type .Number is remembered as Double by default.
        Double doubleArgument = argument.asDouble();
        // We have to provide ExportedModule with proper Number value
        if (expectedArgumentClass == byte.class || expectedArgumentClass == Byte.class) {
          return doubleArgument.byteValue();
        } else if (expectedArgumentClass == short.class || expectedArgumentClass == Short.class) {
          return doubleArgument.shortValue();
        } else if (expectedArgumentClass == int.class || expectedArgumentClass == Integer.class) {
          return doubleArgument.intValue();
        } else if (expectedArgumentClass == float.class || expectedArgumentClass == Float.class) {
          return doubleArgument.floatValue();
        } else if (expectedArgumentClass == long.class || expectedArgumentClass == Long.class) {
          return doubleArgument.longValue();
        } else {
          return doubleArgument;
        }
      case Boolean:
        return argument.asBoolean();
      case Null:
        return null;
      default:
        // JS argument is not null, however we can't recognize the type.
        throw new RuntimeException(
            "Don't know how to convert React Native argument of type " + argument.getType() + " to native."
        );
    }
  }
}
