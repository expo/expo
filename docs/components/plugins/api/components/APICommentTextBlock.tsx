import { mergeClasses } from '@expo/styleguide';
import { CodeSquare01Icon } from '@expo/styleguide-icons/outline/CodeSquare01Icon';
import type { ReactNode } from 'react';
import ReactMarkdown from 'react-markdown';
import { type Components } from 'react-markdown';
import remarkGfm from 'remark-gfm';
import remarkSupsub from 'remark-supersub';
import { InlineHelp } from 'ui/components/InlineHelp';

import { Tag } from '~/ui/components/Tag/Tag';
import { CALLOUT } from '~/ui/components/Text';

import { APIBoxSectionHeader } from './APIBoxSectionHeader';
import { CommentData } from '../APIDataTypes';
import {
  getAllTagData,
  getCommentContent,
  getTagData,
  mdComponents,
  parseCommentContent,
} from '../APISectionUtils';
import { ELEMENT_SPACING, STYLES_SECONDARY } from '../styles';
import { APISectionPlatformTags } from './APISectionPlatformTags';

type Props = {
  comment?: CommentData;
  components?: Components;
  beforeContent?: ReactNode;
  afterContent?: ReactNode;
  includePlatforms?: boolean;
  includeSpacing?: boolean;
  inlineHeaders?: boolean;
  emptyCommentFallback?: string;
};

export const APICommentTextBlock = ({
  comment,
  beforeContent,
  afterContent,
  includePlatforms = true,
  includeSpacing = true,
  inlineHeaders = false,
  emptyCommentFallback,
}: Props) => {
  const content = comment?.summary ? getCommentContent(comment.summary) : undefined;

  if (emptyCommentFallback && !content?.length) {
    return <span className="text-quaternary">{emptyCommentFallback}</span>;
  }

  const paramTags = content ? getParamTags(content) : undefined;
  const parsedContent = (
    <ReactMarkdown components={mdComponents} remarkPlugins={[remarkGfm, remarkSupsub]}>
      {parseCommentContent(paramTags ? content?.replaceAll(PARAM_TAGS_REGEX, '') : content)}
    </ReactMarkdown>
  );

  const examples = getAllTagData('example', comment);
  const exampleContent = examples?.map((example, index) => {
    const exampleText = getCommentContent(example.content ?? example.name);
    const isMultiline = /[\n\r]/.test(exampleText);

    return (
      <div
        key={'example-' + index}
        className={mergeClasses(
          ELEMENT_SPACING,
          !isMultiline && 'flex items-center gap-1.5',
          'last:[&>*]:!mb-0'
        )}>
        {inlineHeaders ? (
          <CALLOUT
            className={mergeClasses(
              'my-1.5 flex flex-row items-center gap-1.5 font-medium text-tertiary',
              !isMultiline && 'my-0'
            )}>
            <CodeSquare01Icon className="icon-sm -mt-px text-icon-tertiary" />
            Example
          </CALLOUT>
        ) : (
          <APIBoxSectionHeader text="Example" className="-mx-4 mb-3 mt-1" Icon={CodeSquare01Icon} />
        )}
        <ReactMarkdown components={mdComponents} remarkPlugins={[remarkGfm, remarkSupsub]}>
          {exampleText}
        </ReactMarkdown>
      </div>
    );
  });

  const see = getTagData('see', comment);
  const seeContent = see && (
    <InlineHelp
      className={mergeClasses('shadow-none', `!${ELEMENT_SPACING}`)}
      size="sm"
      type="info-light">
      <ReactMarkdown components={mdComponents} remarkPlugins={[remarkGfm, remarkSupsub]}>
        {`**See:** ` + getCommentContent(see.content)}
      </ReactMarkdown>
    </InlineHelp>
  );

  const hasPlatforms = (getAllTagData('platform', comment)?.length || 0) > 0;

  return (
    <div className={mergeClasses(includeSpacing && 'px-4 [table_&]:!mb-0 [table_&]:px-0')}>
      {includePlatforms && hasPlatforms && (
        <APISectionPlatformTags
          comment={comment}
          prefix={emptyCommentFallback ? 'Only for:' : undefined}
        />
      )}
      {paramTags && (
        <>
          <span className={STYLES_SECONDARY}>Only for:&ensp;</span>
          {paramTags.map(tag => (
            <Tag key={tag} name={tag.split('-')[1]} />
          ))}
        </>
      )}
      {beforeContent}
      {parsedContent}
      {afterContent}
      {afterContent && !exampleContent && <br />}
      {seeContent}
      {exampleContent}
    </div>
  );
};

const PARAM_TAGS_REGEX = /@tag-\S*/g;

function getParamTags(shortText?: string) {
  if (!shortText?.includes('@tag-')) {
    return undefined;
  }
  return Array.from(shortText.matchAll(PARAM_TAGS_REGEX), match => match[0]);
}
