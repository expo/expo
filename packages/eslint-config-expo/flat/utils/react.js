const pluginReact = require('eslint-plugin-react');
const pluginReactHooks = require('eslint-plugin-react-hooks');

module.exports = [
  {
    plugins: {
      react: pluginReact,
      'react-hooks': pluginReactHooks,
    },

    languageOptions: {
      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
    },

    settings: {
      react: {
        version: 'detect',
      },
    },

    rules: {
      ...pluginReactHooks.configs.recommended.rules,
      'react/display-name': 'warn',
      'react/jsx-no-duplicate-props': 'error',
      'react/jsx-no-undef': 'error',
      'react/jsx-uses-react': 'warn',
      'react/jsx-uses-vars': 'warn',
      'react/no-danger-with-children': 'warn',
      'react/no-deprecated': 'warn',
      'react/no-direct-mutation-state': 'warn',

      'react/no-string-refs': [
        'warn',
        {
          noTemplateLiterals: true,
        },
      ],

      'react/no-this-in-sfc': 'warn',
      'react/no-unknown-property': 'warn',
      'react/require-render-return': 'warn',
    },
  },
];
