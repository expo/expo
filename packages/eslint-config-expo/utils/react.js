module.exports = {
  parserOptions: { ecmaFeatures: { jsx: true } },
  plugins: ['react'],
  extends: ['plugin:react-hooks/recommended'],
  rules: {
    'react/display-name': 'warn',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-undef': 'error',
    'react/jsx-uses-react': 'warn',
    'react/jsx-uses-vars': 'warn',
    'react/no-danger-with-children': 'warn',
    'react/no-deprecated': 'warn',
    'react/no-direct-mutation-state': 'warn',
    'react/no-string-refs': ['warn', { noTemplateLiterals: true }],
    'react/no-this-in-sfc': 'warn',
    'react/no-unknown-property': 'warn',
    'react/require-render-return': 'warn',
  },
  settings: {
    react: { version: 'detect' },
  },
};
