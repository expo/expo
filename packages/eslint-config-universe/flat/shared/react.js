const { defineConfig } = require('eslint/config');
const react = require('eslint-plugin-react');
const hooksPlugin = require('eslint-plugin-react-hooks');

module.exports = defineConfig([
  {
    plugins: {
      react,
      'react-hooks': hooksPlugin,
    },

    languageOptions: {
      // TODO: Const is a reserved word
      // ecmaVersion: 5,
      // sourceType: 'script',

      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
    },

    settings: {
      react: {
        version: 'detect',
      },
    },

    rules: {
      ...hooksPlugin.configs.recommended.rules,
      'react/jsx-boolean-value': ['warn', 'never'],

      'react/jsx-closing-bracket-location': [
        'warn',
        {
          nonEmpty: 'after-props',
          selfClosing: 'tag-aligned',
        },
      ],

      'react/jsx-curly-brace-presence': ['warn', 'never'],

      'react/jsx-curly-spacing': [
        'warn',
        {
          when: 'never',
        },
      ],

      'react/jsx-equals-spacing': ['warn', 'never'],
      'react/jsx-first-prop-new-line': ['warn', 'multiline'],
      'react/jsx-fragments': ['warn', 'syntax'],
      'react/jsx-indent': ['warn', 2],
      'react/jsx-indent-props': ['warn', 2],

      'react/jsx-no-bind': [
        'warn',
        {
          allowArrowFunctions: true,
          allowFunctions: true,
        },
      ],

      'react/jsx-no-duplicate-props': 'error',
      'react/jsx-no-undef': 'error',
      'react/jsx-props-no-multi-spaces': 'warn',
      'react/jsx-tag-spacing': 'warn',
      'react/jsx-uses-react': 'warn',
      'react/jsx-uses-vars': 'warn',

      'react/jsx-wrap-multilines': [
        'warn',
        {
          declaration: false,
          assignment: false,
          return: true,
          arrow: true,
        },
      ],

      'react/no-access-state-in-setstate': 'warn',
      'react/no-did-mount-set-state': 'warn',
      'react/no-did-update-set-state': 'warn',
      'react/no-direct-mutation-state': 'warn',
      'react/no-redundant-should-component-update': 'warn',
      'react/no-this-in-sfc': 'warn',
      'react/no-unknown-property': 'warn',
      'react/no-will-update-set-state': 'warn',
      'react/require-render-return': 'warn',
      'react/self-closing-comp': 'warn',
      'react-hooks/rules-of-hooks': 'error',
      'react-hooks/exhaustive-deps': 'off',
    },
  },
]);
