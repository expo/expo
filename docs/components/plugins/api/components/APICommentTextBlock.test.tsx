import { render } from '@testing-library/react';

import { APICommentTextBlock } from './APICommentTextBlock';
import type { CommentData } from '../APIDataTypes';

describe(APICommentTextBlock, () => {
  test('no comment', () => {
    const { container } = render(<APICommentTextBlock comment={undefined} />);
    expect(container).toMatchSnapshot();
  });

  test('basic comment', () => {
    const comment: CommentData = {
      summary: [{ kind: 'text', text: 'This is the basic comment.' }],
    };

    const { container } = render(<APICommentTextBlock comment={comment} />);
    expect(container).toMatchSnapshot();
  });

  test('comment with example', () => {
    const comment: CommentData = {
      summary: [
        {
          kind: 'text',
          text: 'Gets the referrer URL of the installed app with the [',
        },
        {
          kind: 'code',
          text: '`Install Referrer API`',
        },
        {
          kind: 'text',
          text: '](https://developer.android.com/google/play/installreferrer)\nfrom the Google Play Store. In practice, the referrer URL may not be a complete, absolute URL.',
        },
      ],
      blockTags: [
        {
          tag: '@returns',
          content: [
            {
              kind: 'text',
              text: 'A ',
            },
            {
              kind: 'code',
              text: '`Promise`',
            },
            {
              kind: 'text',
              text: ' that fulfills with a ',
            },
            {
              kind: 'code',
              text: '`string`',
            },
            {
              kind: 'text',
              text: ' of the referrer URL of the installed app.',
            },
          ],
        },
        {
          tag: '@example',
          content: [
            {
              kind: 'code',
              text: '```ts\nawait Application.getInstallReferrerAsync();\n// "utm_source=google-play&utm_medium=organic"\n```',
            },
          ],
        },
        {
          tag: '@platform',
          content: [
            {
              kind: 'text',
              text: 'android',
            },
          ],
        },
      ],
    };

    const { container } = render(<APICommentTextBlock comment={comment} />);
    expect(container).toMatchSnapshot();
  });
});
