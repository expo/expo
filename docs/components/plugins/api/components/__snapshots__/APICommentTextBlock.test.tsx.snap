// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APICommentTextBlock basic comment 1`] = `
<div>
  <div
    class="px-4 [table_&]:!mb-0 [table_&]:px-0"
  >
    <p
      class="text-default font-normal text-[14px] leading-[1.5715] tracking-[-0.006rem] mb-2.5 [table_&]:last:mb-0"
      data-text="true"
    >
      This is the basic comment.
    </p>
  </div>
</div>
`;

exports[`APICommentTextBlock comment with example 1`] = `
<div>
  <div
    class="px-4 [table_&]:!mb-0 [table_&]:px-0"
  >
    <div
      class="mb-3 flex flex-row items-start [table_&]:mb-2.5"
    >
      <span
        class="text-default font-normal text-[14px] leading-[1.5715] tracking-[-0.006rem] inline-flex items-center"
        data-text="true"
      >
        <div
          class="mr-2 inline-flex min-h-[21px] select-none items-center gap-1 rounded-full border px-[7px] py-0.5 last:mr-0 [table_&]:mt-0 [table_&]:px-1.5 [table_&]:py-0.5 [h3_&]:last-of-type:mr-0 [h4_&]:last-of-type:mr-0 bg-palette-green3 text-palette-green12 border-palette-green4"
        >
          <svg
            class="translate-z icon-2xs mt-[-0.5px] shrink-0 text-palette-green12 opacity-80"
            fill="none"
            role="img"
            viewBox="0 0 24 24"
          >
            <g
              id="android-custom-icon"
            >
              <path
                d="M17.0628 15.073C16.5576 15.073 16.1467 14.6605 16.1467 14.1537C16.1467 13.6469 16.5576 13.2347 17.0628 13.2347C17.5679 13.2347 17.9788 13.6469 17.9788 14.1537C17.9788 14.6605 17.5679 15.073 17.0628 15.073ZM6.9372 15.073C6.43204 15.073 6.02116 14.6605 6.02116 14.1537C6.02116 13.6469 6.43204 13.2347 6.9372 13.2347C7.44234 13.2347 7.85325 13.6469 7.85325 14.1537C7.85325 14.6605 7.44234 15.073 6.9372 15.073ZM17.3913 9.53705L19.2222 6.35598C19.3272 6.17313 19.2647 5.93959 19.0828 5.834C18.9008 5.72871 18.6677 5.79135 18.5624 5.97389L16.7087 9.19523C15.291 8.54619 13.6989 8.18468 12 8.18468C10.3011 8.18468 8.70893 8.54619 7.29131 9.19523L5.43751 5.97389C5.33226 5.79135 5.09916 5.72871 4.9172 5.834C4.73524 5.93959 4.67251 6.17313 4.77776 6.35598L6.60865 9.53705C3.46479 11.2524 1.31456 14.4454 1 18.2177H23C22.6851 14.4454 20.5349 11.2524 17.3913 9.53705Z"
                fill="currentColor"
                id="Vector"
              />
            </g>
          </svg>
          <span
            class="whitespace-nowrap !text-3xs font-normal !leading-none"
          >
            Android
          </span>
        </div>
      </span>
    </div>
    <p
      class="text-default font-normal text-[14px] leading-[1.5715] tracking-[-0.006rem] mb-2.5 [table_&]:last:mb-0"
      data-text="true"
    >
      Gets the referrer URL of the installed app with the 
      <a
        class="cursor-pointer decoration-0 hocus:opacity-80 font-normal text-link visited:text-link hocus:underline [&_code]:hocus:underline [&_b]:text-link [&_code]:text-link [&_em]:text-link [&_i]:text-link [&_span]:text-link [&_strong]:text-link"
        href="https://developer.android.com/google/play/installreferrer"
        rel="noopener noreferrer"
        target="_blank"
      >
        <code
          class="text-default text-[13px] font-normal leading-[130%] tracking-[-0.003rem] inline-block rounded-md border border-secondary bg-subtle px-1 break-words !inline py-0"
          data-text="true"
        >
          Install Referrer API
        </code>
      </a>
      
from the Google Play Store. In practice, the referrer URL may not be a complete, absolute URL.
    </p>
    <div
      class="mb-2.5 [table_&]:last:mb-0 last:[&>*]:!mb-0"
    >
      <p
        class="tracking-[-0.006rem] flex border-y border-palette-gray4 bg-subtle px-4 py-2 text-2xs font-medium text-tertiary -mx-4 mb-3 mt-1"
        data-text="true"
      >
        <span
          class="tracking-[-0.006rem] flex flex-row items-center gap-2 text-2xs font-medium text-tertiary"
          data-text="true"
        >
          <svg
            class="translate-z icon-sm text-icon-secondary"
            fill="none"
            role="img"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <g
              id="code-square-01-outline-icon"
            >
              <path
                d="M14.5 15L17.5 12L14.5 9M9.5 9L6.5 12L9.5 15M7.8 21H16.2C17.8802 21 18.7202 21 19.362 20.673C19.9265 20.3854 20.3854 19.9265 20.673 19.362C21 18.7202 21 17.8802 21 16.2V7.8C21 6.11984 21 5.27976 20.673 4.63803C20.3854 4.07354 19.9265 3.6146 19.362 3.32698C18.7202 3 17.8802 3 16.2 3H7.8C6.11984 3 5.27976 3 4.63803 3.32698C4.07354 3.6146 3.6146 4.07354 3.32698 4.63803C3 5.27976 3 6.11984 3 7.8V16.2C3 17.8802 3 18.7202 3.32698 19.362C3.6146 19.9265 4.07354 20.3854 4.63803 20.673C5.27976 21 6.11984 21 7.8 21Z"
                id="Icon"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </g>
          </svg>
          Example
        </span>
      </p>
      <pre
        class="relative my-4 overflow-x-auto whitespace-pre rounded-md border border-secondary bg-subtle p-4 [p+&]:mt-0"
        data-text="true"
      >
        <code
          class="text-2xs text-default"
        >
          <span
            class="token keyword"
          >
            await
          </span>
           Application
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token function"
          >
            getInstallReferrerAsync
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
          

          <span
            class="token comment"
          >
            // "utm_source=google-play&utm_medium=organic"
          </span>
          

        </code>
      </pre>
    </div>
  </div>
</div>
`;

exports[`APICommentTextBlock no comment 1`] = `
<div>
  <div
    class="px-4 [table_&]:!mb-0 [table_&]:px-0"
  />
</div>
`;
